Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp

```diff
@@ -194,10 +196,33 @@ void RemoteGPU::createCompositorIntegration(WebGPUIdentifier identifier)
     ASSERT(m_backing);
 
     auto compositorIntegration = m_backing->createCompositorIntegration();
-    auto remoteCompositorIntegration = RemoteCompositorIntegration::create(compositorIntegration, m_objectHeap, *m_streamConnection, identifier);
+    auto remoteCompositorIntegration = RemoteCompositorIntegration::create(compositorIntegration, m_objectHeap, *m_streamConnection, *this, identifier);
     m_objectHeap->addObject(identifier, remoteCompositorIntegration);
 }
 
+void RemoteGPU::paintNativeImageToImageBuffer(WebCore::NativeImage& nativeImage, WebCore::RenderingResourceIdentifier imageBufferIdentifier)
+{
+    assertIsCurrent(workQueue());
+    Lock lock;
+    Condition conditionVariable;
+    bool isFinished = false;
```

Comment:
This idiom here looks exactly like `wtf::BinarySemaphore`. I think you can replace most of the manual code with:

```
    assertIsCurrent(workQueue());
    BinarySemaphore semaphore;
    
    auto* gpu = m_backing.get();
    m_renderingBackend->dispatch([&]() mutable {
        if (auto imageBuffer = m_renderingBackend->imageBuffer(imageBufferIdentifier)) {
            gpu->paintToCanvas(nativeImage, imageBuffer->backendSize(), imageBuffer->context());
            imageBuffer->flushDrawingContext();
        }
        semaphore.signal();
    });
    semaphore.wait();
}

Reply:
Oh thanks! Yes I should clean that up. I just copied it from the WebGL implementation and removed the parts I didn’t need